In [2]:
import pandas as pd

df = pd.read_excel(r'C:\Users\LENOVO\Downloads\Flight_Information.xlsx')
print(df)

     FlightID FlightNumber    Airline  Destination     Status
0        1001       FL1102  Airline D      Houston    On Time
1        1002       FL1435  Airline B      Chicago    On Time
2        1003       FL1860  Airline A     New York  Cancelled
3        1004       FL1270  Airline C      Chicago    Delayed
4        1005       FL1106  Airline C     New York    Delayed
..        ...          ...        ...          ...        ...
195      1196       FL1863  Airline C     New York  Cancelled
196      1197       FL1742  Airline A     New York    Delayed
197      1198       FL1240  Airline D  Los Angeles  Cancelled
198      1199       FL1563  Airline D      Chicago    On Time
199      1200       FL1095  Airline B      Houston    On Time

[200 rows x 5 columns]


In [4]:
df.isnull().sum()

FlightID        0
FlightNumber    0
Airline         0
Destination     0
Status          0
dtype: int64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   FlightID      200 non-null    int64 
 1   FlightNumber  200 non-null    object
 2   Airline       200 non-null    object
 3   Destination   200 non-null    object
 4   Status        200 non-null    object
dtypes: int64(1), object(4)
memory usage: 7.9+ KB


##  Is there a relationship between the Airline and Flight Status? ##

``` null hypothesis - There's no relation between flight status and airline```
``` alternative hypothesis - There is relation between flight status and airline```

In [10]:
from scipy.stats import chi2_contingency

table = pd.crosstab(df['Status'], df['Airline']) # pd.crosstab creates frquency table
print(table)

Airline    Airline A  Airline B  Airline C  Airline D
Status                                               
Cancelled         18          9         15         18
Delayed           15         10         15         18
On Time           15         22         19         26


In [23]:
#now running chi square test

chi2, p_value, dof, expected = chi2_contingency(table)

print(f'Chi2 value is {chi2}, p value is {p_value}')

Chi2 value is 5.026692061852325, p value is 0.5403942477512074


Here p value is 0.54 which is above 0.05, thus it fails to reject null hypothesis. We can say there is no relation between between flight status and airline.

Note — When to Use chi2_contingency() and chisquare()

chi2_contingency() → Used to test if there is a relationship between two categorical variables.
Example: Is there a relationship between Airline and Flight Status?
→ Needs a 2D contingency table (rows × columns).

chisquare() → Used to check if observed data fits an expected distribution (goodness of fit test). It needs one categorical column.
Example: Do the observed flight delays match the expected distribution?
→ Needs observed vs expected frequencies (1D array).

In [12]:
df

FlightID FlightNumber    Airline  Destination     Status
0        1001       FL1102  Airline D      Houston    On Time
1        1002       FL1435  Airline B      Chicago    On Time
2        1003       FL1860  Airline A     New York  Cancelled
3        1004       FL1270  Airline C      Chicago    Delayed
4        1005       FL1106  Airline C     New York    Delayed
..        ...          ...        ...          ...        ...
195      1196       FL1863  Airline C     New York  Cancelled
196      1197       FL1742  Airline A     New York    Delayed
197      1198       FL1240  Airline D  Los Angeles  Cancelled
198      1199       FL1563  Airline D      Chicago    On Time
199      1200       FL1095  Airline B      Houston    On Time

[200 rows x 5 columns]

## Check if all airlines have equal number of flights

``` null hypothesis - all flights have equal no. of flights```
```` alternative hypothesis - all flights does not have equal no. of flights

In [16]:
count_flight = df['Airline'].value_counts().sort_values()
print(count_flight)



Airline
Airline B    41
Airline A    48
Airline C    49
Airline D    62
Name: count, dtype: int64


In [17]:
total = sum(count_flight)
print(total)

200


In [18]:
len(count_flight)

4

In [24]:
expected = [total/ len(count_flight)]* len(count_flight)
print(expected)

[50.0, 50.0, 50.0, 50.0]


In [25]:
from scipy.stats import chisquare

#performing chisquare test

chi2, p_value = chisquare(count_flight, expected)

print(f'chi2 is {chi2} and p_value is {p_value}')

chi2 is 4.6 and p_value is 0.2035420816793649


Here p_value is 0.20 is above 0.05, it fails to reject null hypothesis. So more or less all flights have equal no. of flights

# Anova Test - It needs numerical columns

Lets run anova test on uber dataset. To check if droploaction affects customer ratings
null hypothesis = no diffrenece
alternative = diffrence is there

In [26]:
import pandas as pd

uber_df = pd.read_csv(r'C:\Users\LENOVO\Desktop\Python\Internshala\Py_Stats_Z_Test _VS_T_Test_Uber_Dataset\archive (1)\ncr_ride_bookings.csv')
print(uber_df)

              Date      Time    Booking ID   Booking Status   Customer ID  \
0       2024-03-23  12:29:38  "CNR5884300"  No Driver Found  "CID1982111"   
1       2024-11-29  18:01:39  "CNR1326809"       Incomplete  "CID4604802"   
2       2024-08-23  08:56:10  "CNR8494506"        Completed  "CID9202816"   
3       2024-10-21  17:17:25  "CNR8906825"        Completed  "CID2610914"   
4       2024-09-16  22:08:00  "CNR1950162"        Completed  "CID9933542"   
...            ...       ...           ...              ...           ...   
149995  2024-11-11  19:34:01  "CNR6500631"        Completed  "CID4337371"   
149996  2024-11-24  15:55:09  "CNR2468611"        Completed  "CID2325623"   
149997  2024-09-18  10:55:15  "CNR6358306"        Completed  "CID9925486"   
149998  2024-10-05  07:53:34  "CNR3030099"        Completed  "CID9415487"   
149999  2024-03-10  15:38:03  "CNR3447390"        Completed  "CID4108667"   

         Vehicle Type         Pickup Location      Drop Location  Avg VTAT 

In [27]:
uber_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 21 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Date                               150000 non-null  object 
 1   Time                               150000 non-null  object 
 2   Booking ID                         150000 non-null  object 
 3   Booking Status                     150000 non-null  object 
 4   Customer ID                        150000 non-null  object 
 5   Vehicle Type                       150000 non-null  object 
 6   Pickup Location                    150000 non-null  object 
 7   Drop Location                      150000 non-null  object 
 8   Avg VTAT                           139500 non-null  float64
 9   Avg CTAT                           102000 non-null  float64
 10  Cancelled Rides by Customer        10500 non-null   float64
 11  Reason for cancelling by Customer  1050

In [30]:
subset = uber_df[['Drop Location', 'Customer Rating']]

In [31]:
subset['Customer Rating'].fillna(subset['Customer Rating'].mean(), inplace=True)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_4880\551544222.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  subset['Customer Rating'].fillna(subset['Customer Rating'].mean(), inplace=True)
C:\Users\LENOVO\AppData\Local\Temp\ipykernel_4880\551544222.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  subset['Customer Rating'].fillna(subset[

In [32]:
subset

Drop Location  Customer Rating
0                 Jhilmil         4.404584
1       Gurgaon Sector 56         4.404584
2           Malviya Nagar         4.900000
3                Inderlok         5.000000
4             Khan Market         4.300000
...                   ...              ...
149995           Ghitorni         4.100000
149996         Akshardham         5.000000
149997           Jor Bagh         4.400000
149998         Saidulajab         3.700000
149999  Gurgaon Sector 29         4.900000

[150000 rows x 2 columns]

In [33]:
subset.isnull().any(axis=1)

0         False
1         False
2         False
3         False
4         False
          ...  
149995    False
149996    False
149997    False
149998    False
149999    False
Length: 150000, dtype: bool

In [38]:
## creating groups

groups = [g['Customer Rating'] for _, g in subset.groupby('Drop Location')]
print(groups)

[11        4.404584
231       4.600000
331       3.800000
680       4.404584
682       4.404584
            ...   
149293    4.500000
149329    4.200000
149567    4.200000
149723    4.300000
149810    4.700000
Name: Customer Rating, Length: 850, dtype: float64, 9         4.404584
696       4.500000
834       4.300000
883       5.000000
1061      4.300000
            ...   
149068    4.900000
149387    4.404584
149519    4.404584
149705    5.000000
149868    4.404584
Name: Customer Rating, Length: 845, dtype: float64, 60        3.300000
143       4.700000
529       4.600000
688       4.404584
1048      3.800000
            ...   
149334    3.000000
149620    4.404584
149676    4.400000
149856    4.404584
149996    5.000000
Name: Customer Rating, Length: 856, dtype: float64, 234       5.000000
518       4.900000
676       4.300000
1090      3.800000
1185      4.404584
            ...   
148438    5.000000
148509    5.000000
148880    4.404584
149175    4.200000
149504    4.404584
Name: C

In [39]:
# running anova test

from scipy.stats import f_oneway

f_stat, p_value = f_oneway(*groups)

print(f'f_stat is {f_stat} and p_value is {p_value}')

f_stat is 0.9907593048161998 and p_value is 0.5203556983699706


**Here the p_value is 0.52 ahive is above 0.05, it fails to reject null hypothesis. So we can say drop loaction doesnt affect customer's rating much.**

| Thing                      | Meaning                                                         |
| -------------------------- | --------------------------------------------------------------- |
| `groupby('Drop Location')` | Breaks the big DataFrame into smaller groups based on locations |
| `g['Customer Rating']`     | Selects only the rating column from each small group            |
| `[ ... for _, g in ... ]`  | Creates a list of ratings for all locations                     |
| `groups`                   | A list of all ratings grouped by location                       |
| Numbers like 11, 231, 331  | Original DataFrame row indexes                                  |
| Locations                  | Stored separately as group names (`keys()`)                     |
